pkg/kubectl/rollback.go

```diff
@@ -119,113 +115,86 @@ func (r *DeploymentRollbacker) Rollback(obj runtime.Object, updatedAnnotations m
 		return "", fmt.Errorf("failed to retrieve Deployment %s: %v", name, err)
 	}
 
+	templateForRevision, latestTemplate, err := deploymentRevision(deployment, r.c, toRevision)
+	if err != nil {
+		return "", err
+	}
 	if dryRun {
-		return simpleDryRun(deployment, r.c, toRevision)
+		return printTemplate(templateForRevision)
 	}
 	if deployment.Spec.Paused {
 		return "", fmt.Errorf("you cannot rollback a paused deployment; resume it first with 'kubectl rollout resume deployment/%s' and try again", name)
 	}
-	deploymentRollback := &extensionsv1beta1.DeploymentRollback{
-		Name:               name,
-		UpdatedAnnotations: updatedAnnotations,
-		RollbackTo: extensionsv1beta1.RollbackConfig{
-			Revision: toRevision,
-		},
-	}
-	result := ""
 
-	// Get current events
-	events, err := r.c.CoreV1().Events(namespace).List(metav1.ListOptions{})
-	if err != nil {
-		return result, err
-	}
-	// Do the rollback
-	// TODO: This is DEPRECATED. It should be updated. DaemonSets and StatefulSets implement rollback by
-	// patching using history (ControllerRevision data). Deployments should probably also implement
-	// rollback using a patch.
-	if err := r.c.ExtensionsV1beta1().Deployments(namespace).Rollback(deploymentRollback); err != nil {
-		return result, err
+	// Skip if the revision already matches current Deployment
+	if apiequality.Semantic.DeepEqual(templateForRevision, latestTemplate) {
+		return fmt.Sprintf("%s (current template already matches revision %d)", rollbackSkipped, toRevision), nil
 	}
-	// Watch for the changes of events
-	watch, err := r.c.CoreV1().Events(namespace).Watch(metav1.ListOptions{Watch: true, ResourceVersion: events.ResourceVersion})
+
+	// make patch to restore
+	patchType, patch, err := getDeploymentPatch(templateForRevision)
 	if err != nil {
-		return result, err
+		return "", fmt.Errorf("failed restoring revision %d: %v", toRevision, err)
 	}
-	result = watchRollbackEvent(watch)
-	return result, err
-}
 
-// watchRollbackEvent watches for rollback events and returns rollback result
-func watchRollbackEvent(w watch.Interface) string {
-	signals := make(chan os.Signal, 1)
-	signal.Notify(signals, os.Interrupt, os.Kill, syscall.SIGTERM)
-	for {
-		select {
-		case event, ok := <-w.ResultChan():
-			if !ok {
-				return ""
-			}
-			obj, ok := event.Object.(*corev1.Event)
-			if !ok {
-				w.Stop()
-				return ""
-			}
-			isRollback, result := isRollbackEvent(obj)
-			if isRollback {
-				w.Stop()
-				return result
-			}
-		case <-signals:
-			w.Stop()
-		}
+	// Restore revision
+	if _, err = r.c.AppsV1().Deployments(namespace).Patch(name, patchType, patch); err != nil {
```

Comment:
We need to do an update instead of a patch. A rollback should change the template back to a previous revision, instead of merging changes. 

We also need to handle some special logic such as not copying pod-template-hash labels, not copying template annotations, and reset deployment annotations. Basically this:
https://github.com/kubernetes/kubernetes/blob/553a169d256154f7e601a40c1f617d1d2f219cf0/pkg/controller/deployment/rollback.go#L79-L91

Reply:
> A rollback should change the template back to a previous revision, instead of merging changes.

this is replacing, not merging (the generated patch is doing a replace operation on the template)

> We also need to handle some special logic such as not copying pod-template-hash labels, not copying template annotations, and reset deployment annotations.

will update